File: py/735.asteroid-collision.py
#
# @lc app=leetcode id=735 lang=python3
#
# [735] Asteroid Collision
#

# @lc code=start
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        aster_stack = []
        for n in asteroids:
            while aster_stack and n < 0 < aster_stack[-1]:
                if aster_stack[-1] < -n:
                    aster_stack.pop()
                elif aster_stack[-1] == -n:
                    aster_stack.pop()
                    break
                else:
                    break
            else:
                aster_stack.append(n)
        return aster_stack
# @lc code=end

File: cpp/735.asteroid-collision.cpp
/*
 * @lc app=leetcode id=735 lang=cpp
 *
 * [735] Asteroid Collision
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<int> asteroidCollision(vector<int> &asteroids)
    {
        vector<int> aster_stack;
        for (auto n : asteroids)
        {
            while (!aster_stack.empty() && aster_stack.back() > 0 && n < 0)
            {
                if (aster_stack.back() < -n)
                {
                    aster_stack.pop_back();
                }
                else
                {
                    break;
                }
            }
            if (aster_stack.empty())
            {
                aster_stack.push_back(n);
            }
            else
            {
                if (n > 0 || aster_stack.back() < 0)
                {
                    aster_stack.push_back(n);
                }
                else if (aster_stack.back() == -n)
                {
                    aster_stack.pop_back();
                }
            }
        }
        return aster_stack;
    }
};
// @lc code=end
